File: misc/LPTM/merge_particles.py
################################################################################
#                                                                              # 
#   Merge raw particle output from the Lagrangian particle module in UCLA-LES  #
#   Requires numpy, netCDF4, pylab, which are all provided on MPI systems      #
#   and certain DKRZ systems (at least Lizard) via module python/2.7-ve0       #
#    - Bart van Stratum, Aug 2012                                              #
#                                                                              #
#   Command-line input: script asks for the variables to be merged             #   
#   'all' puts all variables (x,y,z,u,v,w,t,tv,rt,rl) into one NetCDF file     #
#                                                                              #
#   TODO: add units, description, etc. to NetCDF output                        #  
#                                                                              #
################################################################################

import numpy;
from netCDF4 import Dataset
from pylab import *
import time
import sys

# USER INPUT:
pref   = 'arm'       # Path to particle output 
nfx    = 8           # Number of cores in x-direction
nfy    = 3           # Number of cores in y-direction

# No need to change below
# ----------------------------------------------
vars = raw_input('variables (e.g. x,y,u,t,.. or all): ').split(',')   # commandline input
if(vars[0] == 'all'):
  vars = (['x','y','z','u','v','w','t','tv','rt','rl'])
print 'Merging: ', vars

globalstart = time.time()
# 1. Find total number of particles: 
np = 0
for i in range(nfx):
  for j in range(nfy):
    filein = pref + '.particles.' + str(i).zfill(4) + str(j).zfill(4) + '.nc'
    ncin   = Dataset(filein, 'r')
    np    += size(ncin.variables["particles"])

    if(i==0 and j==0):  # check if variables present
      for var in vars:
        try:
          varin = ncin.variables[var.strip()]
        except KeyError as e:
          sys.exit("One or more variables not found, stopping..")

print 'Found ', np, ' particles!'

# 2. Create new NetCDF
fileout = pref + '.particles.' + '_'.join(vars) + '.nc'
ncout   = Dataset(fileout, 'w', format='NETCDF4')
pdim    = ncout.createDimension('particles', np)
tdim    = ncout.createDimension('time'      ,None)
print 'output file = ',fileout

x  = ncout.createVariable('x', 'f4',('time','particles',)) if 'x'  in vars else 1
y  = ncout.createVariable('y', 'f4',('time','particles',)) if 'y'  in vars else 1
z  = ncout.createVariable('z', 'f4',('time','particles',)) if 'z'  in vars else 1
u  = ncout.createVariable('u', 'f4',('time','particles',)) if 'u'  in vars else 1
v  = ncout.createVariable('v', 'f4',('time','particles',)) if 'v'  in vars else 1
w  = ncout.createVariable('w', 'f4',('time','particles',)) if 'w'  in vars else 1
t  = ncout.createVariable('t', 'f4',('time','particles',)) if 't'  in vars else 1
tv = ncout.createVariable('tv','f4',('time','particles',)) if 'tv' in vars else 1
rt = ncout.createVariable('rt','f4',('time','particles',)) if 'rt' in vars else 1
rl = ncout.createVariable('rl','f4',('time','particles',)) if 'rl' in vars else 1

# 3. Read in file-by-file and write to new NetCDF
loc = 0
for i in range(nfx):
  for j in range(nfy):
    start = time.time()
    filein = pref + '.particles.' + str(i).zfill(4) + str(j).zfill(4) + '.nc'
    ncin   = Dataset(filein, 'r')
    npl    = size(ncin.variables["particles"])

    if 'x' in vars:
      x[:,loc:loc+npl]  = ncin.variables["x"][:,:] 
    if 'y' in vars:
      y[:,loc:loc+npl]  = ncin.variables["y"][:,:] 
    if 'z' in vars:
      z[:,loc:loc+npl]  = ncin.variables["z"][:,:] 
    if 'u' in vars:
      u[:,loc:loc+npl]  = ncin.variables["u"][:,:] 
    if 'v' in vars:
      v[:,loc:loc+npl]  = ncin.variables["v"][:,:] 
    if 'w' in vars:
      w[:,loc:loc+npl]  = ncin.variables["w"][:,:] 
    if 't' in vars:
      t[:,loc:loc+npl]  = ncin.variables["t"][:,:] 
    if 'tv' in vars:
      tv[:,loc:loc+npl] = ncin.variables["tv"][:,:]
    if 'rt' in vars:
      rt[:,loc:loc+npl] = ncin.variables["rt"][:,:]
    if 'rl' in vars:
      rl[:,loc:loc+npl] = ncin.variables["rl"][:,:]

    ncin.close()
    loc += npl
    print 'Processed: ',filein, ' in ', time.time()-start, 's' 

ncout.close()
print 'Total user time: ', time.time()-globalstart, 's' 

File: misc/LPTM/merge_particles.f90
!
! Merge raw particle data from UCLA-LES
! Command line input: 'prefix var nxprocs nyprocs'
! e.g: './merge_particles dir/rico x 8 8'
! Only one variable can be provided, allows for poor man's 
! parallelization using a number of serial batch jobs
!

program mergeparticles
  use netcdf
  implicit none
  
  character(100)                    :: pref,cnx,cny,path,var
  integer                           :: nx,ny
  integer                           :: nargs
  character(100)                    :: varname,path_out
  integer                           :: ncid_in,varid_in
  integer                           :: ncid_out,pdim_out,tdim_out,pid_out,tid_out,varid_out,start_in(2),start_out(2),count(2)
  integer                           :: npin=0,nptot=0,ntime=0,ntin=0
  integer                           :: i,j,t,loc
  character(8)                      :: procxy
  real, allocatable, dimension(:)   :: var_read
  
  real                              :: etime
  real*4                            :: elapsed(2)
  real                              :: total
  
  !-- MAIN CALLS
  ! Get input user 
  nargs = command_argument_count()
  if(nargs .lt. 4) then 
    print*,'provide -prefix varflg nxprox nyprocs- as arguments (see source code), STOPPING'
    stop
  else
    call get_command_argument(1,pref)
    call get_command_argument(2,var)
    call get_command_argument(3,cnx)
    call get_command_argument(4,cny)
    read (cnx,*) nx
    read (cny,*) ny
  end if
  
  ! Loop through files to read number of times/particles 
  do i=1,nx
    do j=1,ny
      write(procxy,'(i4.4,i4.4)') i-1,j-1
      path = trim(pref)//".particles."//procxy//".nc"
      call check(nf90_open(path,nf90_nowrite,ncid_in))
      call check(nf90_inq_dimid(ncid_in, "particles", varid_in))
      call check(nf90_inquire_dimension(ncid_in, varid_in, varname, npin))
      call check(nf90_inq_dimid(ncid_in, "time", varid_in))
      call check(nf90_inquire_dimension(ncid_in, varid_in, varname, ntin))
      call check(nf90_close(ncid_in))
      if(ntin > ntime) ntime = ntin
      nptot = nptot + npin 
    end do
  end do
  print*,'found ', nptot, 'particles, ', ntime, 'time steps'
 
  ! Create new NetCDF file
  print*,'creating new NetCDF file' 
  path_out = trim(pref)//".particles."//trim(var)//".nc"
  call check(nf90_create(path_out,nf90_clobber,ncid_out))
  ! Create dimensions
  call check(nf90_def_dim(ncid_out,'particles',nptot,pdim_out))
  call check(nf90_def_dim(ncid_out,'time',ntime,tdim_out))
  ! Create variables
  !call check(nf90_def_var(ncid_out,'particles',nf90_double,(/pdim_out/),pid_out))
  call check(nf90_def_var(ncid_out,'time',nf90_double,(/tdim_out/),tid_out))
  call check(nf90_def_var(ncid_out,trim(var),nf90_float,(/pdim_out,tdim_out/),varid_out))
  ! Leave define mode
  call check(nf90_enddef(ncid_out))

  ! Copy data
  print *,'merging files' 
  loc = 1
  do i=1,nx
    do j=1,ny
      write(procxy,'(i4.4,i4.4)') i-1,j-1
      path = trim(pref)//".particles."//procxy//".nc"
      call check(nf90_open(path,nf90_nowrite,ncid_in))
      call check(nf90_inq_dimid(ncid_in, "particles", varid_in))
      call check(nf90_inquire_dimension(ncid_in, varid_in, varname, npin))
  
      allocate(var_read(npin))
  
      call check(nf90_inq_varid(ncid_in, trim(var), varid_in))
      count  = (/npin,1/)
      do t=1,ntin
        start_in  = (/1,t/)
        start_out = (/loc,t/) 
        call check(nf90_get_var(ncid_in,  varid_in,  var_read, start=start_in, count=count))
        call check(nf90_put_var(ncid_out, varid_out, var_read, start=start_out, count=count)) 
      end do
  
      loc = loc + npin
      deallocate(var_read)

      print*,i,j

      if(i .eq. nx .and. j .eq. ny) then
        allocate(var_read(nptot))
        call check(nf90_inq_varid(ncid_in, 'time', varid_in))
        call check(nf90_get_var(ncid_in,  varid_in, var_read, start=(/1/), count=(/ntime/)))
        call check(nf90_put_var(ncid_out, tid_out, var_read, start=(/1/), count=(/ntime/))) 
        deallocate(var_read)
      end if

      call check(nf90_close(ncid_in))

      print*,'Processed ',trim(path),(i-1)*nx+j,'/',nx*ny 
    end do
  end do
  
  call check(nf90_close(ncid_out))

contains

  subroutine check(status)
    integer, intent (in) :: status
    if(status /= nf90_noerr) then 
      print *, trim(nf90_strerror(status))
      stop "Stopped"
    end if
  end subroutine check

end program mergeparticles
